keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+
+The device plugin support added in Kubernetes 1.8 makes it easy for vendors to dynamically export their resources through a plugin API without changing Kubernetes core code. Taking a further step, this document proposes a new resource abstraction API, ResourceClass, that can be used to describe, manage and consume such vendor specific and metadata rich resources in simple and portable ways.
+
+## Use Stories
+### As a cluster operator:
+- Nodes in my cluster has GPU HW from different generations. I want to classify GPU nodes into one of the three categories, silver, gold and platinum depending upon the launch timeline of the GPU family eg: Kepler K20, K80, Pascal P40, P100, Volta V100. I want to charge each of the three categories differently. I want to offer my clients 3 GPU rates/classes to choose from.<br/>
+**Motivation:** As time progresses in a cluster lifecycle, new advanced, high performance, expensive variants of GPUs gets added to the cluster nodes. At the same time older variants also co-exist. There are workloads which strictly wants latest GPUs and also there are workloads which are fine with older GPUs. But since there is a wide range of types, it will be hard to manage and confusing at the same time to have granularity at each GPU type. Grouping into few broad categories will be convenient to manage.<br/>
+**Can this be solved without resource classes:** A unique taint can be used to represent a category like silver. Nodes can be tainted accordingly depending upon the type of GPUs availability. User pods can use tolerations to steer workloads to the appropriate nodes. But problem is how to restrict a user pod from not using the toleration that it should not be using?<br/>
+**How Resource classes can solve this:** I, operator/admin, creates three resource classes: GPU-Platinum, GPU-Gold, GPU-Silver. Now since resource classes are quota controlled, end-user will be able to request resource classes only if quota is allocated.
+
+- I want a mechanism where it is possible to offer a group of devices, which are co-located on a single node and shares a common property, as a single resource that can be requested in pod container spec. Example, N GPU units interconnected by NVLink or N cpu cores on same NUMA node.<br/>
+**Motivation:** Increased performance because of local access. Local access also helps better use of cache<br/>
+**How Resource classes can solve this:**  Property/attribute which forms the grouping can be advertised in the device attributes and then a resource can be created to form a grouped super-resource based on that property.<br/>
+**Can this be solved without resource classes:** No
+
+- I want to have quota control on the devices at the granularity of device properties. For example, I want to have a separate quota for ECC enabled GPUs. I want a specific user to not let use more than ‘N’ number of ECC enabled GPUs overall at namespace level.<br/>
```

Comment:
I'm not convinced quota control on resource attributes is a real use case.
Especially ECC, since it is a property that when it is present on a GPU can be dynamically enabled or disabled.

Is that something that is used at Red Hat @vikaschoudhary16 ?

Reply:
Just speaking from GKE production experience, quota control is important for enterprise customers. @vikaschoudhary16 may also speak from Openshift experience.

Reply:
Certainly quota control is something that we are interested to have in Openshift. @derekwaynecarr @jeremyeder

Reply:
Ah sorry for a moment I forgot that GPU type is an attribute.

What I meant to say here is that the user story "I want this group to only have 10 ECC enabled GPUs" isn't something that we expect to see and seems a bit far fetched.

Especially If we're going to define Resource Classes as a billable unit then I'd expect a lot more the user story to be "I want this group to have 10 K80 GPUs and 20 V100 GPUs".

Reply:
i agree that the likely quota use case is 10 K80 GPUs and 20 V100 GPUs

Reply:
+1. Let's solve the quota per device problem first.